### PR TITLE
Clean the implementation and API of autorewrite databases

### PR DESCRIPTION
**Pull request by ppedrot**
This should be behaviour preserving, except maybe for slight differences in performance (and involuntarily introduced bugs). This is a prerequisite cleanip before actually changing the semantics of autorewrite hint registering to properly implement #16332.

This PR is essentially moving code around and hiding implementation details under proper APIs.

Overlays:
- https://github.com/mattam82/Coq-Equations/pull/501

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
There are real anomalies in there, in the end passing an additional evarmap in the API is probably not that backwards compatible.

**Comment by ppedrot** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
The bench died for unknown reasons btw

**Comment by ppedrot** [comment]
@SkySkimmer I killed it because the CI was broken. Now that it seems to be working properly, let's relaunch it. @coqbot bench

**Comment by ppedrot** [comment]
This looks ready to me.

**Comment by ppedrot** [comment]
This needs an assignee, @SkySkimmer maybe?

**Review by SkySkimmer — APPROVED** [review]

File: tactics/dnet.ml [hunk 1]
```diff
@@ -134,39 +128,6 @@ struct
 (*     | (Some s, _ | _, Some s) -> s *)
 (*     | _ -> raise Not_found *)
```
Comment (SkySkimmer) [inline tactics/dnet.ml:129]: This commented code should be removed too.

Comment (ppedrot) [reply tactics/dnet.ml:129]: I didn't touch it, it was there before and was not moved. It's not strictly part of the code I'm fixing either...

Comment (SkySkimmer) [reply tactics/dnet.ml:129]: It goes together with the fold_pattern stuff deleted just below

Comment (ppedrot) [reply tactics/dnet.ml:129]: OK, will do.

Comment (ppedrot) [reply tactics/dnet.ml]: Done.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**